jax/experimental/mosaic/gpu/core.py

```diff
@@ -511,9 +511,11 @@ def _launch(
     profiler_start = (smem_bytes + align - 1) & ~(align - 1)
     smem_bytes = profiler_start + profiler_spec.smem_bytes(block=block)
 
-  # TODO(cperivol): Query the shared memory size programmatically.
-  if smem_bytes > 228 * 1024:
-    raise ValueError(f"Mosaic GPU kernel exceeds available shared memory {smem_bytes=} > 228000")
+  # assumes all available devices have the same amount of shared memory
+  max_smem_bytes = jax.local_devices()[0].shared_memory_per_block_optin
```

Comment:
Can you use `getattr` to optionally try to query this and otherwise just fall back to what we had before? Unfortunately we still need to support e.g. compiling MGPU kernels on CPU for export, and no CUDA device will be attached in that case